daemons/dss-sync-sfn/app.py

```diff
@@ -0,0 +1,306 @@
+            bucket = resources.s3.Bucket(event_record["s3"]["bucket"]["name"])
+            obj = bucket.Object(unquote(event_record["s3"]["object"]["key"]))
+            if bucket.name != source_replica.bucket:
+                logger.error("Received S3 event for bucket %s with no configured replica", bucket.name)
+                continue
+
+            for dest_replica in Config.get_replication_destinations(source_replica):
+                if exists(dest_replica, obj.key):
+                    logger.error("Key %s already exists in %s, skipping sync", obj.key, dest_replica)
+                    continue
+                exec_name = bucket.name + "/" + obj.key + ":" + source_replica.name + ":" + dest_replica.name
+                exec_input = dict(source_replica=source_replica.name,
+                                  dest_replica=dest_replica.name,
+                                  source_key=obj.key,
+                                  source_obj_metadata=event_record["s3"]["object"])
+                executions[exec_name] = app.state_machine.start_execution(**exec_input)["executionArn"]
+    return executions
+
+# This entry point is for external events forwarded by dss-gs-event-relay (or other event sources) through SQS.
+@app.sqs_queue_subscriber("dss-sync-" + os.environ["DSS_DEPLOYMENT_STAGE"])
+def launch_from_forwarded_event(event, context):
+    executions = {}
+    for event_record in event["Records"]:
+        message = json.loads(event_record["body"])
+        if message["selfLink"].startswith("https://www.googleapis.com/storage"):
+            source_replica = Replica.gcp
+            source_key = message["name"]
+            bucket = source_replica.bucket
+            for dest_replica in Config.get_replication_destinations(source_replica):
+                if exists(dest_replica, source_key):
+                    logger.info("Key %s already exists in %s, skipping sync", source_key, dest_replica)
+                    continue
+                exec_name = bucket + "/" + message["name"] + ":" + source_replica.name + ":" + dest_replica.name
+                exec_input = dict(source_replica=source_replica.name,
+                                  dest_replica=dest_replica.name,
+                                  source_key=message["name"],
+                                  source_obj_metadata=message)
+                executions[exec_name] = app.state_machine.start_execution(**exec_input)["executionArn"]
+        else:
+            raise NotImplementedError()
+    return executions
+
+retry_config = [
+    {
+        "ErrorEquals": ["States.TaskFailed"],
+        "IntervalSeconds": 5,
+        "MaxAttempts": 5,
+        "BackoffRate": 1.5
+    },
+    {
+        "ErrorEquals": ["States.Timeout"],
+        "IntervalSeconds": 30,
+        "MaxAttempts": 3,
+        "BackoffRate": 1.5
+    },
+    {
+        "ErrorEquals": ["States.Permissions"],
+        "MaxAttempts": 0
+    }
+]
+
+sfn = {
```

Comment:
The chart (gif) is extremely helpful for understanding `sfn` logic. So hooray for including it :)

Reply:
Thanks. I wanted to include an svg (in the AWS Console, this is rendered as an svg), but Github broke inline svg rendering 😢